src/clients/arclib/arcget.1.in
.\" -*- nroff -*-
.TH ARCGET 1 "@DATE@" "NorduGrid ARC @VERSION@" "NorduGrid Users Manual"

.SH NAME

arcget \- ARC Get

.SH DESCRIPTION

The
.B arcget
command is used for retrieving the results from a job.

.SH SYNOPSIS

.B arcget [options] [job ...]

.SH OPTIONS

.IP "\fB-a\fR, \fB--all\fR"
all jobs
.IP "\fB-j\fR, \fB--joblist\fR=\fIfilename\fR"
the file storing information about active jobs (default ~/.arc/jobs.xml)
.IP "\fB-i\fR, \fB--jobids-from-file\fR=\fIfilename\fR"
a file containing a list of jobIDs
.IP "\fB-c\fR, \fB--cluster\fR=\fIname\fR"
select one or more computing elements: \fIname\fR can be an alias for a single CE, a group of CEs or a URL
.IP "\fB-s\fR, \fB--status\fR=\fIstatusstr\fR"
only select jobs whose status is statusstr
.IP "\fB-D\fR, \fB--dir\fR=\fIdirname\fR"
download directory (the job directory will be created in this directory)
.IP "\fB-J\fR, \fB--usejobname\fR"
use the jobname instead of the short ID as the job directory name
.IP "\fB-k\fR, \fB--keep\fR"
keep files on the remote cluster (do not clean)
.IP "\fB-f\fR, \fB--force\fR"
force download (overwrite existing job directory)
.IP "\fB-P\fR, \fB--listplugins\fR"
list the available plugins
.IP "\fB-t\fR, \fB--timeout\fR=\fIseconds\fR"
timeout in seconds (default 20)
.IP "\fB-z\fR, \fB--conffile\fR=\fIfilename\fR"
configuration file (default ~/.arc/client.conf)
.IP "\fB-d\fR, \fB--debug\fR=\fIdebuglevel\fR"
FATAL, ERROR, WARNING, INFO, VERBOSE or DEBUG
.IP "\fB-v\fR, \fB--version\fR"
print version information
.IP "\fB-?\fR, \fB--help\fR"
print help
.LP

.SH ARGUMENTS

.IP "\fBjob\fR ..."
list of jobids and/or jobnames
.LP

.SH EXTENDED DESCRIPTION

The
.B arcget
command downloads the results after a job has completed on an grid
enabled computing resource. Only the results of jobs that have
finished can be downloaded. The job can be referred to either by the
jobid that was returned by
.BR arcsub (1)
at submission time or by its jobname if the job description that was
submitted contained a jobname attribute.

More than one jobid and/or jobname can be given. If several jobs were
submitted with the same jobname the results of all those jobs are
downloaded. If the
.B --joblist
option is used the list of jobs is read from a file with the
specified filename. By specifying the
.B --all
option, the results of all active jobs are downloaded.

The
.B --cluster
option can be used to select or reject jobs at specific clusters. See
.BR arcsub (1)
for a discussion of the format of arguments to this option. The
.B --status
option can be used to select jobs in a specific state. These options
can be repeated several times.

For each job that is downloaded a subdirectory will be created in the
download directory that will contain the downloaded files.

If the download was successful the job will be removed from the remote
cluster unless the
.B --keep
option was specified.

.SH FILES

.TP
.B ~/.arc/client.conf
Some options can be given default values by specifying them in the
ARC client configuration file. By using the
.B --conffile
option a different configuration file can be used than the default.

.TP
.B ~/.arc/jobs.xml
This a local list of the user's active jobs. When a job is
successfully submitted it is added to this list and when it is removed
from the remote cluster it is removed from this list. This list is
used as the list of all active jobs when the user specifies the
.B --all
option to the various NorduGrid ARC user interface commands. By using the
.B --joblist
option a different file can be used than the default.

.SH ENVIRONMENT VARIABLES

.TP
.B X509_USER_PROXY
The location of the user's grid proxy file. Shouldn't be set unless
the proxy is in a non-standard location.

.TP
.B ARC_LOCATION
The location where ARC is installed can be specified by this variable. If not
specified the install location will be determined from the path to the
command being executed, and if this fails a WARNING will be given stating the
location which will be used.

.TP
.B ARC_PLUGIN_PATH
The location of ARC plugins can be specified by this variable. Multiple
locations can be specified by separating them by : (; in Windows). The
default location is \fB$ARC_LOCATION\fR/lib/arc (\\ in Windows).

.SH COPYRIGHT
APACHE LICENSE Version 2.0

.SH AUTHOR

Mattias Ellert <mattias.ellert@fysast.uu.se>

.SH SEE ALSO

.BR arccat (1),
.BR arcclean (1),
.BR arccp (1),
.BR arcinfo (1),
.BR arckill (1),
.BR arcls (1),
.BR arcmigrate (1),
.BR arcmkdir (1),
.BR arcproxy (1),
.BR arcrenew (1),
.BR arcresub (1),
.BR arcresume (1),
.BR arcrm (1),
.BR arcstat (1),
.BR arcsub (1),
.BR arcsync (1),
.BR arctest (1)



NEWS
no news good news :)


src/clients/arclib/arcget.cpp
// -*- indent-tabs-mode: nil -*-

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <iostream>
#include <list>
#include <string>

#include <arc/ArcLocation.h>
#include <arc/IString.h>
#include <arc/Logger.h>
#include <arc/UserConfig.h>
#include <arc/client/JobSupervisor.h>

#include "utils.h"

int RUNMAIN(arcget)(int argc, char **argv) {

  setlocale(LC_ALL, "");

  Arc::Logger logger(Arc::Logger::getRootLogger(), "arcget");
  Arc::LogStream logcerr(std::cerr);
  logcerr.setFormat(Arc::ShortFormat);
  Arc::Logger::getRootLogger().addDestination(logcerr);
  Arc::Logger::getRootLogger().setThreshold(Arc::WARNING);

  Arc::ArcLocation::Init(argv[0]);

  ClientOptions opt(ClientOptions::CO_GET,
                    istring("[job ...]"),
                    istring("The arcget command is used for "
                            "retrieving the results from a job."));

  std::list<std::string> jobidentifiers = opt.Parse(argc, argv);

  if (opt.showversion) {
    std::cout << Arc::IString("%s version %s", "arcget", VERSION)
              << std::endl;
    return 0;
  }

  // If debug is specified as argument, it should be set before loading the configuration.
  if (!opt.debug.empty())
    Arc::Logger::getRootLogger().setThreshold(Arc::string_to_level(opt.debug));

  if (opt.show_plugins) {
    std::list<std::string> types;
    types.push_back("HED:JobControllerPlugin");
    showplugins("arcget", types, logger);
    return 0;
  }

  Arc::UserConfig usercfg(opt.conffile, opt.joblist);
  if (!usercfg) {
    logger.msg(Arc::ERROR, "Failed configuration initialization");
    return 1;
  }

  if (opt.debug.empty() && !usercfg.Verbosity().empty())
    Arc::Logger::getRootLogger().setThreshold(Arc::string_to_level(usercfg.Verbosity()));

  if (opt.downloaddir.empty()) {
    if (!usercfg.JobDownloadDirectory().empty()) {
      opt.downloaddir = usercfg.JobDownloadDirectory();
      logger.msg(Arc::INFO, "Job download directory from user configuration file: %s ", opt.downloaddir);
    }
    else {
      logger.msg(Arc::INFO, "Job download directory will be created in present working directory. ");
    }
  }
  else {
    logger.msg(Arc::INFO, "Job download directory: %s ", opt.downloaddir);
  }

  for (std::list<std::string>::const_iterator it = opt.jobidinfiles.begin(); it != opt.jobidinfiles.end(); it++) {
    if (!Arc::Job::ReadJobIDsFromFile(*it, jobidentifiers)) {
      logger.msg(Arc::WARNING, "Cannot read specified jobid file: %s", *it);
    }
  }

  if (opt.timeout > 0)
    usercfg.Timeout(opt.timeout);

  if ((!opt.joblist.empty() || !opt.status.empty()) && jobidentifiers.empty() && opt.clusters.empty())
    opt.all = true;

  if (jobidentifiers.empty() && opt.clusters.empty() && !opt.all) {
    logger.msg(Arc::ERROR, "No jobs given");
    return 1;
  }

  std::list<std::string> selectedURLs = getSelectedURLsFromUserConfigAndCommandLine(usercfg, opt.clusters);
  std::list<std::string> rejectManagementURLs = getRejectManagementURLsFromUserConfigAndCommandLine(usercfg, opt.rejectmanagement);

  std::list<Arc::Job> jobs;
  if (!Arc::Job::ReadJobsFromFile(usercfg.JobListFile(), jobs, jobidentifiers, opt.all, selectedURLs, rejectManagementURLs)) {
    logger.msg(Arc::ERROR, "Unable to read job information from file (%s)", usercfg.JobListFile());
    return 1;
  }

  for (std::list<std::string>::const_iterator itJIdentifier = jobidentifiers.begin();
       itJIdentifier != jobidentifiers.end(); ++itJIdentifier) {
    std::cout << Arc::IString("Warning: Job not found in job list: %s", *itJIdentifier) << std::endl;
  }

  Arc::JobSupervisor jobmaster(usercfg, jobs);

  jobmaster.Update();
  jobmaster.SelectValid();
  if (!opt.status.empty()) {
    jobmaster.SelectByStatus(opt.status);
  }

  if (jobmaster.GetSelectedJobs().empty()) {
    std::cout << Arc::IString("No jobs") << std::endl;
    return 1;
  }

  std::list<std::string> downloaddirectories;
  int retval = (int)!jobmaster.Retrieve(opt.downloaddir, opt.usejobname, opt.forcedownload, downloaddirectories);

  for (std::list<std::string>::const_iterator it = downloaddirectories.begin();
       it != downloaddirectories.end(); ++it) {
    std::cout << Arc::IString("Results stored at: %s", *it) << std::endl;
  }

  unsigned int retrieved_num = jobmaster.GetIDsProcessed().size();
  unsigned int notretrieved_num = jobmaster.GetIDsNotProcessed().size();
  unsigned int cleaned_num = 0;

  if (!opt.keep) {
    std::list<Arc::URL> retrieved = jobmaster.GetIDsProcessed();
    // No need to clean selection because retrieved is subset of selected
    jobmaster.SelectByID(retrieved);
    if(!jobmaster.Clean()) {
      std::cout << Arc::IString("Warning: Some jobs were not removed from server") << std::endl;
      std::cout << Arc::IString("         Use arclean to remove retrieved jobs from job list", usercfg.JobListFile()) << std::endl;
      retval = 1;
    }
    cleaned_num = jobmaster.GetIDsProcessed().size();

    if (!Arc::Job::RemoveJobsFromFile(usercfg.JobListFile(), jobmaster.GetIDsProcessed())) {
      std::cout << Arc::IString("Warning: Failed to lock job list file %s", usercfg.JobListFile()) << std::endl;
      std::cout << Arc::IString("         Use arclean to remove retrieved jobs from job list", usercfg.JobListFile()) << std::endl;
      retval = 1;
    }

    std::cout << Arc::IString("Jobs processed: %d, successfully retrieved: %d, successfully cleaned: %d", retrieved_num+notretrieved_num, retrieved_num, cleaned_num) << std::endl;

  } else {

    std::cout << Arc::IString("Jobs processed: %d, successfully retrieved: %d", retrieved_num+notretrieved_num, retrieved_num) << std::endl;

  }

  return retval;
}
